plonky3/src/circuit_builder.rs

```diff
@@ -0,0 +1,202 @@
+
+use std::collections::BTreeMap;
+
+use p3_air::{Air, AirBuilder, BaseAir};
+use p3_field::AbstractField;
+use p3_matrix::{dense::RowMajorMatrix, MatrixRowSlices};
+use powdr_ast::analyzed::{
+    AlgebraicBinaryOperator, AlgebraicExpression, AlgebraicUnaryOperator, Analyzed, IdentityKind,
+    PolynomialType,
+};
+use powdr_executor::witgen::WitgenCallback;
+use powdr_number::{FieldElement, LargeInt};
+
+pub type Val = p3_goldilocks::Goldilocks;
+
+pub(crate) struct PowdrCircuit<'a, T> {
+    /// The analyzed PIL
+    analyzed: &'a Analyzed<T>,
+    /// The number of committed polynomials, computed from `analyzed` and cached
+    commitment_count: usize,
+    /// The value of the fixed columns
+    fixed: &'a [(String, Vec<T>)],
+    /// The value of the witness columns, if set
+    witness: Option<&'a [(String, Vec<T>)]>,
+    /// Column name and index of the public cells
+    publics: Vec<(String, usize)>,
+    /// Callback to augment the witness in the later stages
+    _witgen_callback: Option<WitgenCallback<T>>,
+}
+
+pub fn cast_to_goldilocks<T: FieldElement>(v: T) -> Val {
+    Val::from_canonical_u64(v.to_integer().try_into_u64().unwrap())
+}
+
+impl<'a, T: FieldElement> PowdrCircuit<'a, T> {
+    pub(crate) fn new(analyzed: &'a Analyzed<T>, fixed: &'a [(String, Vec<T>)]) -> Self {
+        let mut publics = analyzed
+            .public_declarations
+            .values()
+            .map(|public_declaration| {
+                let witness_name = public_declaration.referenced_poly_name();
+                let witness_offset = public_declaration.index as usize;
+                (witness_name, witness_offset)
+            })
+            .collect::<Vec<_>>();
+        // Sort, so that the order is deterministic
+        publics.sort();
+
+        Self {
+            analyzed,
+            commitment_count: analyzed.commitment_count(),
+            fixed,
+            witness: None,
+            publics,
+            _witgen_callback: None,
+        }
+    }
+
+    fn witness(&self) -> &'a [(String, Vec<T>)] {
+        self.witness.as_ref().unwrap()
+    }
+
+    pub(crate) fn with_witness(self, witness: &'a [(String, Vec<T>)]) -> Self {
+        Self {
+            witness: Some(witness),
+            ..self
+        }
+    }
+
+    pub(crate) fn with_witgen_callback(self, witgen_callback: WitgenCallback<T>) -> Self {
+        Self {
+            _witgen_callback: Some(witgen_callback),
+            ..self
+        }
+    }
+
+    /// Computes the instance column from the witness
+    pub(crate) fn instance_column(&self) -> Vec<Val> {
+        let witness = self
+            .witness
+            .as_ref()
+            .expect("Witness needs to be set")
+            .iter()
+            .map(|(name, values)| (name, values))
+            .collect::<BTreeMap<_, _>>();
+
+        self.publics
+            .iter()
+            .map(|(col_name, i)| cast_to_goldilocks(witness.get(col_name).unwrap()[*i]))
+            .collect()
```

Comment:
Do you actually constrain them somewhere? If not, I think we should assert that the publics are empty for now.

Reply:
Good point, I did not implement that logic in `eval`. Removing as it can be a separate PR.